core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -133,44 +134,49 @@ protected WritePrimaryResult shardOperationOnPrimary(BulkShardRequest request, I
-                    response = operationResult.hasFailure() ? null
+                    Engine.IndexResult indexResult = executeIndexRequestOnPrimary(indexRequest, primary, mappingUpdatedAction);
+                    operationResult = indexResult;
+                    response = indexResult.hasFailure() ? null
                             : new IndexResponse(primary.shardId(), indexRequest.type(), indexRequest.id(),
-                                operationResult.getVersion(), ((Engine.IndexResult) operationResult).isCreated());
+                                indexResult.getVersion(), indexResult.isCreated());
                     break;
                 case UPDATE:
                     UpdateResultHolder updateResultHolder = executeUpdateRequest(((UpdateRequest) itemRequest),
                             primary, metaData, request, requestIndex);
                     operationResult = updateResultHolder.operationResult;
                     response = updateResultHolder.response;
+                    replicaRequest = updateResultHolder.replicaRequest;
                     break;
                 case DELETE:
                     final DeleteRequest deleteRequest = (DeleteRequest) itemRequest;
-                    operationResult = executeDeleteRequestOnPrimary(deleteRequest, primary);
-                    response = operationResult.hasFailure() ? null :
+                    Engine.DeleteResult deleteResult = executeDeleteRequestOnPrimary(deleteRequest, primary);
+                    operationResult = deleteResult;
+                    response = deleteResult.hasFailure() ? null :
                             new DeleteResponse(request.shardId(), deleteRequest.type(), deleteRequest.id(),
-                                operationResult.getVersion(), ((Engine.DeleteResult) operationResult).isFound());
+                                deleteResult.getVersion(), deleteResult.isFound());
                     break;
                 default: throw new IllegalStateException("unexpected opType [" + itemRequest.opType() + "] found");
             }
             // update the bulk item request because update request execution can mutate the bulk item request
-            BulkItemRequest item = request.items()[requestIndex];
-            if (operationResult == null // in case of a noop update operation
-                    || operationResult.hasFailure() == false) {
-                if (operationResult != null) {
-                    location = locationToSync(location, operationResult.getLocation());
-                } else {
-                    assert response.getResult() == DocWriteResponse.Result.NOOP
-                            : "only noop update can have null operation";
-                }
-                // set update response
-                item.setPrimaryResponse(new BulkItemResponse(item.id(), opType, response));
+            request.items()[requestIndex] = replicaRequest;
```

Comment:
I opened https://github.com/elastic/elasticsearch/issues/21221 for this, I agree we should restore the original request, but it is non-trivial because we update the request in place. 
In reality though, it should not be a problem, because upon retrying the shard bulk request, the shard will just execute the translated request instead of translating the update request again.